File: sdk-libs/ts/src/keypair/pda.ts
import { getAddressEncoder, type Address } from "@solana/kit";

import { type Bytes32, type Bytes64 } from "./bytes.js";
import { P_PDA, PdaRoleExpansion } from "./derivation.js";
import { KeypairError } from "./error.js";
import { NullifierKey } from "./nullifier-key.js";
import { P256PublicKey, ShieldedPublicKey, type Curve, type ViewTag } from "./public-key.js";
import {
  CompressedShieldedAddress,
  ShieldedAddress,
  type ShieldedKeypairLike,
  type ViewingKeyLike,
} from "./shielded.js";
import { type Salt, VIEWING_KEY_ECDH_RAW, ViewingKey } from "./viewing-key.js";

const addressEncoder = getAddressEncoder();

/** A shielded identity owned by a program-derived address. */
export class ShieldedPda implements ShieldedKeypairLike, ViewingKeyLike {
  readonly #pda: Address;
  readonly #nullifier: NullifierKey;
  readonly #viewing: ViewingKey;

  private constructor(pda: Address, nullifier: NullifierKey, viewing: ViewingKey) {
    this.#pda = pda;
    this.#nullifier = nullifier;
    this.#viewing = viewing;
  }

  static fromKeyExchange(pda: Address, own: ViewingKey, counterparty: P256PublicKey): ShieldedPda {
    return ShieldedPda.#expand(pda, own.ecdh(counterparty));
  }

  static fromViewingKey(pda: Address, own: ViewingKey): ShieldedPda {
    return ShieldedPda.#expand(pda, own[VIEWING_KEY_ECDH_RAW](P_PDA));
  }

  static fromParts(pda: Address, nullifier: NullifierKey, viewing: ViewingKey): ShieldedPda {
    return new ShieldedPda(pda, nullifier, viewing);
  }

  static #expand(pda: Address, shared: Bytes32): ShieldedPda {
    const pdaBytes = new Uint8Array(addressEncoder.encode(pda)) as Bytes32;
    const expansion = new PdaRoleExpansion(shared, pdaBytes);
    try {
      return new ShieldedPda(pda, expansion.nullifierKey(), expansion.viewingKey());
    } finally {
      shared.fill(0);
      expansion.destroy();
    }
  }

  pda(): Address {
    return this.#pda;
  }

  signingPublicKey(): ShieldedPublicKey {
    return ShieldedPublicKey.fromPda(new Uint8Array(addressEncoder.encode(this.#pda)) as Bytes32);
  }

  viewingPublicKey(): P256PublicKey {
    return this.#viewing.publicKey();
  }

  viewingKey(): ViewingKey {
    const secret = this.#viewing.secretBytes();
    try {
      return ViewingKey.fromBytes(secret);
    } finally {
      secret.fill(0);
    }
  }

  nullifierKey(): NullifierKey {
    const secret = this.#nullifier.secretBytes();
    try {
      return NullifierKey.fromSecret(secret);
    } finally {
      secret.fill(0);
    }
  }

  curve(): Curve {
    return "pda";
  }

  shieldedAddress(): ShieldedAddress {
    return ShieldedAddress.forPda(this.#pda, this.#nullifier.publicKey(), this.viewingPublicKey());
  }

  nullifierPublicKey(): Bytes32 {
    return this.#nullifier.publicKey();
  }

  ownerHash(): Bytes32 {
    return this.shieldedAddress().ownerHash();
  }

  compressedAddress(): CompressedShieldedAddress {
    return CompressedShieldedAddress.fromParts(this.ownerHash(), this.viewingPublicKey());
  }

  sign(_message: Uint8Array): Bytes64 {
    throw new KeypairError("KEYPAIR_PDA_CANNOT_SIGN");
  }

  nullifier(utxoHash: Bytes32, blinding: Bytes32): Bytes32 {
    return this.#nullifier.nullifier(utxoHash, blinding);
  }

  publicKey(): P256PublicKey {
    return this.viewingPublicKey();
  }

  ecdh(counterparty: P256PublicKey): Bytes32 {
    return this.#viewing.ecdh(counterparty);
  }

  mergeViewTag(mergeCount: bigint): ViewTag {
    return this.#viewing.mergeViewTag(mergeCount);
  }

  recipientBootstrapViewTag(): ViewTag {
    return this.#viewing.recipientBootstrapViewTag();
  }

  transactionViewingKey(firstNullifier: Bytes32): ViewingKey {
    return this.#viewing.transactionViewingKey(firstNullifier);
  }

  encryptSlot(
    recipientPublicKey: P256PublicKey,
    plaintext: Uint8Array,
    salt: Salt,
    slotIndex: number,
  ): Uint8Array {
    return this.#viewing.encryptSlot(recipientPublicKey, plaintext, salt, slotIndex);
  }

  decryptUtxo(
    ciphertext: Uint8Array,
    txViewingPublicKey: P256PublicKey,
    salt: Salt,
    slotIndex: number,
  ): Uint8Array {
    return this.#viewing.decryptUtxo(ciphertext, txViewingPublicKey, salt, slotIndex);
  }

  decryptSlotEphemeral(
    recipientPublicKey: P256PublicKey,
    ciphertext: Uint8Array,
    salt: Salt,
    slotIndex: number,
  ): Uint8Array {
    return this.#viewing.decryptSlotEphemeral(recipientPublicKey, ciphertext, salt, slotIndex);
  }

  encryptVerifiable(
    userViewingPublicKey: P256PublicKey,
    plaintext: Uint8Array,
  ): Readonly<{ ciphertext: Uint8Array; txViewingPublicKey: P256PublicKey }> {
    return this.#viewing.encryptVerifiable(userViewingPublicKey, plaintext);
  }

  decryptVerifiable(txViewingPublicKey: P256PublicKey, ciphertext: Uint8Array): Uint8Array {
    return this.#viewing.decryptVerifiable(txViewingPublicKey, ciphertext);
  }

  destroy(): void {
    this.#nullifier.destroy();
    this.#viewing.destroy();
  }
}

File: sdk-libs/keypair/src/pda.rs
//! Shielded identities owned by a program-derived address
//! (`docs/shielded_pda_design.md`). A PDA is off the Ed25519 curve and holds
//! no private state, so the identity is rooted in an ECDH exchange between
//! viewing keys: both roles expand from that one shared secret, bound to the
//! PDA. Spending needs both the owning program's `invoke_signed` and the
//! nullifier secret; neither alone suffices.

use solana_address::Address;
use zeroize::Zeroizing;

use crate::{
    derivation::{p_pda, PdaRoleExpansion},
    error::KeypairError,
    hash::owner_hash,
    nullifier_key::NullifierKey,
    pubkey::{Curve, P256Pubkey, PublicKey},
    shielded::{CompressedShieldedAddress, ShieldedAddress},
    traits::ShieldedKeypairTrait,
    viewing_key::ViewingKey,
};

/// Not a `ShieldedKeypair`: it holds no signing secret, so that state is
/// unrepresentable rather than checked at run time.
#[derive(Clone)]
pub struct ShieldedPda {
    pda: Address,
    nullifier_key: NullifierKey,
    viewing_key: ViewingKey,
}

impl AsRef<NullifierKey> for ShieldedPda {
    fn as_ref(&self) -> &NullifierKey {
        &self.nullifier_key
    }
}

impl ShieldedPda {
    /// Derives both roles from `ECDH(own, counterparty)`, so either
    /// participant reconstructs the identity from its own viewing key and the
    /// counterparty's viewing public key; nothing is transported. A sole
    /// holder passes its own viewing public key as the counterparty.
    pub fn from_key_exchange(
        pda: Address,
        own: &ViewingKey,
        counterparty: &P256Pubkey,
    ) -> Result<Self, KeypairError> {
        let shared = Zeroizing::new(own.ecdh(counterparty)?);
        let expansion = PdaRoleExpansion::new(&shared, pda.to_bytes());
        Ok(Self {
            pda,
            nullifier_key: expansion.nullifier_key()?,
            viewing_key: expansion.viewing_key()?,
        })
    }

    /// Derives both roles from the holder's viewing key alone via
    /// `ECDH(own, P_pda)`, a committed point with unknown discrete log, so the
    /// identity exists before any counterparty does and its `nullifier_pk` can
    /// be published at account-creation time. The holder set is the same as
    /// the sole-holder exchange; only the holder of `own` derives it.
    pub fn from_viewing_key(pda: Address, own: &ViewingKey) -> Result<Self, KeypairError> {
        let shared = Zeroizing::new(own.ecdh_raw(&p_pda())?);
        let expansion = PdaRoleExpansion::new(&shared, pda.to_bytes());
        Ok(Self {
            pda,
            nullifier_key: expansion.nullifier_key()?,
            viewing_key: expansion.viewing_key()?,
        })
    }

    /// Supplied roles, for the flows the exchange cannot express: three or
    /// more parties that must each build spend proofs, or a protocol that
    /// fixes the values itself. A PDA has no signing key, so nothing can
    /// re-derive a supplied role.
    pub fn with_viewing_key(
        pda: Address,
        nullifier_key: NullifierKey,
        viewing_key: ViewingKey,
    ) -> Self {
        Self {
            pda,
            nullifier_key,
            viewing_key,
        }
    }

    pub fn pda(&self) -> &Address {
        &self.pda
    }

    pub fn signing_pubkey(&self) -> PublicKey {
        PublicKey::from_pda(&self.pda)
    }

    pub fn viewing_key(&self) -> &ViewingKey {
        &self.viewing_key
    }

    pub fn viewing_pubkey(&self) -> P256Pubkey {
        self.viewing_key.pubkey()
    }

    pub fn shielded_address(&self) -> Result<ShieldedAddress, KeypairError> {
        Ok(ShieldedAddress {
            signing_pubkey: self.signing_pubkey(),
            nullifier_pubkey: self.nullifier_key.pubkey()?,
            viewing_pubkey: self.viewing_pubkey(),
        })
    }

    /// The published half of the nullifier role: a PDA identity publishes this
    /// value (in the governing account or the registry) so verifiers can bind
    /// the owner hash without holding the secret.
    pub fn nullifier_pubkey(&self) -> Result<[u8; 32], KeypairError> {
        self.nullifier_key.pubkey()
    }

    pub fn owner_hash(&self) -> Result<[u8; 32], KeypairError> {
        owner_hash(&self.signing_pubkey(), &self.nullifier_key.pubkey()?)
    }

    pub fn compressed_address(&self) -> Result<CompressedShieldedAddress, KeypairError> {
        Ok(CompressedShieldedAddress {
            owner_hash: self.owner_hash()?,
            viewing_pubkey: self.viewing_pubkey(),
        })
    }

    pub fn nullifier(
        &self,
        utxo_hash: &[u8; 32],
        blinding: &[u8; 32],
    ) -> Result<[u8; 32], KeypairError> {
        self.nullifier_key.nullifier(utxo_hash, blinding)
    }
}

impl ShieldedKeypairTrait for ShieldedPda {
    fn signing_pubkey(&self) -> PublicKey {
        self.signing_pubkey()
    }

    fn viewing_pubkey(&self) -> P256Pubkey {
        self.viewing_pubkey()
    }

    fn curve(&self) -> Curve {
        Curve::Pda
    }

    fn shielded_address(&self) -> Result<ShieldedAddress, KeypairError> {
        self.shielded_address()
    }

    fn owner_hash(&self) -> Result<[u8; 32], KeypairError> {
        self.owner_hash()
    }

    fn compressed_address(&self) -> Result<CompressedShieldedAddress, KeypairError> {
        self.compressed_address()
    }

    fn sign_message(&self, _message: &[u8]) -> Result<[u8; 64], KeypairError> {
        Err(KeypairError::PdaCannotSign)
    }

    fn sign_hash(&self, _hash: &[u8; 32]) -> Result<[u8; 64], KeypairError> {
        Err(KeypairError::PdaCannotSign)
    }

    fn nullifier(
        &self,
        utxo_hash: &[u8; 32],
        blinding: &[u8; 32],
    ) -> Result<[u8; 32], KeypairError> {
        self.nullifier(utxo_hash, blinding)
    }

    fn nullifier_key(&self) -> NullifierKey {
        self.nullifier_key.clone()
    }
}

// Forwards to the identity's derived viewing key, so encryption and view-tag
// derivation need no special case.
crate::forward_viewing_key_trait!(ShieldedPda => viewing_key);
